src/js/modules/utilities/IOBuf.ts

```diff
@@ -206,19 +226,43 @@ export class IOBuf {
     return iob;
   }
 
+  /**
+   * find the first fragment with free space, otherwise it returns the last
+   * fragment in the list.
+   */
   private getLastFragment = (): Fragment => {
-    const lastFragment = this.fragmentList[this.fragmentList.length - 1];
+    const lastFragment = this.fragmentList[this.currentWriteIndex];
     if (lastFragment) {
       return lastFragment;
     } else {
-      return this.appendNewFragment();
+      return this.appendNewFragment(lastFragment?.getSize());
+    }
+  };
+
+  /**
+   * given a fragment, it tries to find the next fragment in the list , if there
+   * isn't next one, it appends new one.
```

Comment:
given a fragment, it tries to find the next fragment in the list, or appends a new one.*

Reply:
Done